### PR TITLE
API v4 with JWT Authentication

### PR DESCRIPTION
**Pull request by Goutte**
Initial rushes of an implementation of the API using ApiPlatform and Json Web Tokens (JWT).

Of note: 

- Requires a database migration
- Users are now linked to Contributors (users have hats / contributors have impersonators)
- Publishing a Notice in the name of a Contributor is only allowed if the authenticated User is allowed to impersonate the Contributor.
- Doc and sandbox at `/api/v4/docs.html`.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -89,6 +95,17 @@ $ aphp bin/console assets:install web
 $ dMigrate && dLoad
 ```
```
[the remark this answers is not in the document]
Comment (lutangar) [reply CONTRIBUTING.md:97]: For example, this was written during the meeting:
```
# API

> API V3 is left untouched

## Endpoints

The endpoint to publish `Notice` as an authenticated `User` is only available on API V4.
Available endpoints are (will be) described at https://notices.bulles.fr/api/v3/api/v4/docs

## Clients

Client libraries could be generated in any programming languages using the OpenApi definition file.
The definition file is available in `Json` format at: https://notices.bulles.fr/api/v3/api/v4/docs.json

## Authentication


> Authentication system is only available on API V4.

Authentication is implemented as a Json Web Token (JWT).
Currently name login + password 
see...
```

File: src/Entity/Notice.php [hunk 2]
```diff
@@ -4,42 +4,109 @@
...
 
     /**
-     * @var ?string
+     * The visibility of the Notice.  See `NoticeVisibility`
+     * for an enumeration of the allowed values:
+     *   - "public": anyone may view this Notice
+     *   - "private": Notice is only visible to Contributor
+     *   - "archived": ???
+     *   - "draft": Notice is only visible to Contributor, pending publication
+     *   - "question": ???
```
Comment (lutangar) [inline src/Entity/Notice.php:95]: A question asked by a someone on a webpage. A question is not publicly visible.

File: src/Entity/Notice.php [hunk 3]
```diff
@@ -4,42 +4,109 @@
...
      */
     private $id;
 
     /**
-     * @var ?string
+     * The visibility of the Notice.  See `NoticeVisibility`
+     * for an enumeration of the allowed values:
+     *   - "public": anyone may view this Notice
+     *   - "private": Notice is only visible to Contributor
+     *   - "archived": ???
```
Comment (lutangar) [inline src/Entity/Notice.php:93]: A deleted notice because it wasn't relevant anymore.

File: src/Entity/Notice.php [hunk 4]
```diff
@@ -53,6 +120,8 @@ class Notice
      */
     private $matchingContexts;
 
+    // Goutte: Is this still used?  Isn't it in matchingContexts?  Can anyone document?
```
Comment (lutangar) [inline src/Entity/Notice.php:123]: @JalilArfaoui is it?

File: src/Entity/Notice.php (hunk 2 above)
Comment (Goutte) [reply src/Entity/Notice.php:95]: Thanks !

**Comment by JalilArfaoui** [comment]
@lutangar @Goutte I’ve plugged this branch to its own deployment https://v4-notices.bulles.fr/api/v4 so that we can start having functional reviews and feedbacks from @MaartenLMEM  @zhinu @julienadamcom together with the code review

**Review by JalilArfaoui — COMMENTED** [review]
First, thanks for initiating this

I left some comments here and there …

Overall, I’d be curious to see how we would implement a POST `/contributions` like we have on V3 …

Also, this could benefit a bit of clean-up in the commits' history (commits reverted, comments cleaned later, duplicated migration …)

Maybe more important, I’d argue that we don’t need a many-to-many relation between users and contributors, but rather a `contributor` `belongs to` `user`, that may suits the use case better, especially since we have an admin user that can act of all contributors …
What do you think @MaartenLMEM : Do we really need a contributor to belong to multiple user accounts ? 
I know who can do more can do less, but  added complexity has to prove useful …

File: migrations/Version20210430133028.php [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+use Doctrine\Migrations\AbstractMigration;
+
+/**
+ * Add a relation between users and contributors.
+ */
+final class Version20210430133028 extends AbstractMigration
+{
+    public function up(Schema $schema) : void
+    {
+        $this->addSql('CREATE TABLE user_contributor (user_id INT NOT NULL, contributor_id INT NOT NULL, INDEX IDX_73745E06A76ED395 (user_id), INDEX IDX_73745E067A19A357 (contributor_id), PRIMARY KEY(user_id, contributor_id)) DEFAULT CHARACTER SET utf8 COLLATE `utf8_unicode_ci` ENGINE = InnoDB');
```
[the remark this answers is not in the document]
Comment (JalilArfaoui) [reply migrations/Version20210430133028.php:17]: I saw in a latter commit that this has been fixed I guess

**Comment by MaartenLMEM** [comment]
> What do you think @MaartenLMEM : Do we really need a contributor to belong to multiple user accounts ? 
Before we are sustainable , no. And if all is going right, probably not before 2022.

File: config/packages/security.yaml [hunk 6]
```diff
@@ -15,10 +15,12 @@ security:
         dev:
             pattern: ^/(_(profiler|wdt)|css|images|js)/
             security: false
-        api:
+        api_v4_docs:
+            pattern: ^/api/v4(|/|/docs|/docs/|/docs[.][a-z0-9]+)$
+            anonymous: true
```
[the remark this answers is not in the document]
Comment (lutangar) [reply config/packages/security.yaml:20]: > #395 395

**Review by lutangar — APPROVED** [review]
:tada: :tada: :tada: :tada:

**Pull request merged**